File: control/src/mitigations/tls.ts
import { Type, type Static } from '@sinclair/typebox';
import { Value } from '@sinclair/typebox/value';
import type { FastifyInstance, FastifyReply, FastifyRequest } from 'fastify';

/**
 * Schéma TypeBox des params tls-renegotiation-flood. Aligné avec
 * configs/schemas/tls.schema.json et
 * proxy/mitigations/tlsreneg/tlsreneg.go.
 */
export const TLSParamsSchema = Type.Object(
  {
    min_tls_version: Type.Union([Type.Literal('1.2'), Type.Literal('1.3')]),
    handshakes_per_second_per_ip: Type.Number({
      exclusiveMinimum: 0,
      maximum: 1_000_000,
    }),
    burst: Type.Integer({ minimum: 1, maximum: 100_000 }),
  },
  { additionalProperties: false },
);
export type TLSParams = Static<typeof TLSParamsSchema>;

/**
 * Règle de la famille "tls". MVP : seul l'id "tls-renegotiation-flood".
 *
 * Défaut on_error="allow" (fail-open AGENTS.md §3) ; la renégociation
 * cliente est toujours refusée au niveau crypto/tls (non négociable).
 */
export const TLSRuleSchema = Type.Object(
  {
    id: Type.Literal('tls-renegotiation-flood'),
    enabled: Type.Boolean(),
    on_error: Type.Union([Type.Literal('allow'), Type.Literal('deny')]),
    params: TLSParamsSchema,
    notes: Type.Optional(Type.String({ maxLength: 1024 })),
    reason: Type.Optional(Type.String({ maxLength: 256 })),
  },
  { additionalProperties: false },
);
export type TLSRule = Static<typeof TLSRuleSchema>;

export interface TLSStore {
  get(id: string): TLSRule | undefined;
  put(rule: TLSRule): { rev: number };
  list(): { rev: number; rules: TLSRule[] };
}

export function createInMemoryTLSStore(): TLSStore {
  let rev = 0;
  const rules = new Map<string, TLSRule>();
  return {
    get: (id) => rules.get(id),
    put: (rule) => {
      if (!rule.enabled && (rule.reason === undefined || rule.reason.trim() === '')) {
        throw new Error('reason is required when enabled=false');
      }
      rules.set(rule.id, rule);
      rev += 1;
      return { rev };
    },
    list: () => ({ rev, rules: [...rules.values()] }),
  };
}

/**
 * Enregistre les routes /v1/mitigations/tls sur app.
 *
 *   GET  /v1/mitigations/tls         → snapshot
 *   PUT  /v1/mitigations/tls/:id     → upsert avec validation
 *
 * Le push effectif au data plane passe par POST /v1/reload.
 */
export function registerTLSRoutes(app: FastifyInstance, store: TLSStore): void {
  app.route({
    method: 'GET',
    url: '/v1/mitigations/tls',
    schema: {
      response: {
        200: Type.Object(
          {
            rev: Type.Integer({ minimum: 0 }),
            rules: Type.Array(TLSRuleSchema),
          },
          { additionalProperties: false },
        ),
      },
    },
    handler: async () => store.list(),
  });

  app.route({
    method: 'PUT',
    url: '/v1/mitigations/tls/:id',
    schema: {
      params: Type.Object(
        { id: Type.String({ minLength: 1, maxLength: 64 }) },
        { additionalProperties: false },
      ),
      response: {
        200: Type.Object(
          {
            status: Type.Literal('ok'),
            rev: Type.Integer({ minimum: 1 }),
            rule: TLSRuleSchema,
          },
          { additionalProperties: false },
        ),
        400: Type.Object(
          {
            error: Type.Union([
              Type.Literal('id_mismatch'),
              Type.Literal('invalid_rule'),
              Type.Literal('disabled_without_reason'),
            ]),
            details: Type.Array(Type.String()),
          },
          { additionalProperties: false },
        ),
      },
    },
    handler: async (
      req: FastifyRequest<{ Params: { id: string }; Body: unknown }>,
      reply: FastifyReply,
    ) => {
      const body = req.body;
      if (!Value.Check(TLSRuleSchema, body)) {
        const details = [...Value.Errors(TLSRuleSchema, body)].map(
          (e) => `${e.path} ${e.message}`,
        );
        reply.code(400);
        return { error: 'invalid_rule' as const, details };
      }
      if (body.id !== req.params.id) {
        reply.code(400);
        return {
          error: 'id_mismatch' as const,
          details: [`url id=${req.params.id} body id=${body.id}`],
        };
      }
      try {
        const { rev } = store.put(body);
        return { status: 'ok' as const, rev, rule: body };
      } catch (err) {
        reply.code(400);
        const msg = err instanceof Error ? err.message : String(err);
        return { error: 'disabled_without_reason' as const, details: [msg] };
      }
    },
  });
}

File: proxy/internal/adminapi/tls.go
package adminapi

import (
	"encoding/json"
	"net/http"

	"anti-ddos/proxy/mitigations/tlsreneg"
)

// tlsRule : format JSON aligné avec control/src/mitigations/tls.ts
// et configs/schemas/tls.schema.json.
type tlsRule struct {
	ID      string        `json:"id"`
	Enabled bool          `json:"enabled"`
	OnError string        `json:"on_error"`
	Params  tlsRuleParams `json:"params"`
	Reason  string        `json:"reason,omitempty"`
	Notes   string        `json:"notes,omitempty"`
}

type tlsRuleParams struct {
	MinTLSVersion            string  `json:"min_tls_version"`
	HandshakesPerSecondPerIP float64 `json:"handshakes_per_second_per_ip"`
	Burst                    int     `json:"burst"`
}

type tlsPayload struct {
	Rev   int       `json:"rev"`
	Rules []tlsRule `json:"rules"`
}

func tlsHandler(lim *tlsreneg.Limiter) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		switch r.Method {
		case http.MethodGet:
			cfg := lim.Config()
			rule := tlsRule{
				ID:      "tls-renegotiation-flood",
				Enabled: cfg.Enabled,
				OnError: cfg.OnError,
				Params: tlsRuleParams{
					MinTLSVersion:            cfg.MinTLSVersion,
					HandshakesPerSecondPerIP: cfg.HandshakesPerSecondPerIP,
					Burst:                    cfg.Burst,
				},
			}
			writeJSON(w, http.StatusOK, tlsPayload{Rev: 0, Rules: []tlsRule{rule}})
		case http.MethodPut:
			r.Body = http.MaxBytesReader(w, r.Body, 64*1024)
			var payload tlsPayload
			dec := json.NewDecoder(r.Body)
			dec.DisallowUnknownFields()
			if err := dec.Decode(&payload); err != nil {
				writeJSON(w, http.StatusBadRequest, map[string]string{
					"error":   "invalid_body",
					"details": err.Error(),
				})
				return
			}
			var target *tlsRule
			for i := range payload.Rules {
				if payload.Rules[i].ID == "tls-renegotiation-flood" {
					target = &payload.Rules[i]
					break
				}
			}
			if target == nil {
				writeJSON(w, http.StatusOK, map[string]string{"status": "noop"})
				return
			}
			newCfg := tlsreneg.Config{
				Enabled:                  target.Enabled,
				MinTLSVersion:            target.Params.MinTLSVersion,
				HandshakesPerSecondPerIP: target.Params.HandshakesPerSecondPerIP,
				Burst:                    target.Params.Burst,
				OnError:                  target.OnError,
			}
			if err := lim.Update(newCfg); err != nil {
				writeJSON(w, http.StatusBadRequest, map[string]string{
					"error":   "invalid_config",
					"details": err.Error(),
				})
				return
			}
			writeJSON(w, http.StatusOK, map[string]any{
				"status": "applied",
				"rev":    payload.Rev,
			})
		default:
			w.Header().Set("Allow", "GET, PUT")
			w.WriteHeader(http.StatusMethodNotAllowed)
		}
	}
}
